concept/thing/thing_manager.rs

```
@@ -1918,6 +2060,312 @@ impl ThingManager {
+                updated_role_types.extend(role_types.iter());
+            }
+        }
+
+        for (object_type, attribute_types) in modified_owns {
+            let (min, max) = minmax_or!(
+                TypeAPI::chain_types(
+                    object_type,
+                    object_type.get_subtypes_transitive(snapshot, self.type_manager())?.into_iter().cloned()
+                ),
+                unreachable!("Expected at least one object type")
+            );
+            let mut it = self.get_objects_in_range_unordered(snapshot, &(Bound::Included(min), Bound::Included(max)));
+            while let Some(object) = it.next().transpose()? {
+                let updated_attribute_types = out_object_attribute_types.entry(object).or_default();
+                updated_attribute_types.extend(attribute_types.iter());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn collect_modified_schema_capability_cardinalities(
+        &self,
+        snapshot: &impl WritableSnapshot,
+        modified_owns: &mut HashMap<ObjectType, HashSet<AttributeType>>,
+        modified_plays: &mut HashMap<ObjectType, HashSet<RoleType>>,
+        modified_relates: &mut HashMap<RelationType, HashSet<RoleType>>,
+    ) -> Result<(), Box<ConceptReadError>> {
+        // New / deleted capabilities
+
+        for (key, write) in snapshot.iterate_writes_range(&KeyRange::new_within(
+            TypeEdge::build_prefix(Prefix::EdgeOwns),
+            TypeEdge::FIXED_WIDTH_ENCODING,
+        )) {
+            let edge = TypeEdge::decode(Bytes::reference(key.bytes()));
+            let attribute_type = AttributeType::new(edge.to());
+            let updated_attribute_types = modified_owns.entry(ObjectType::new(edge.from())).or_default();
+            match write {
+                Write::Insert { .. } | Write::Put { .. } => {
+                    updated_attribute_types.insert(attribute_type);
```

This is a small optimization to avoid chaining all the supertypes of interface types every time I've mentioned in the long "docs" comment in the `validate` method.

If `owns` is created, it is marked as affected for all the objects of the owner types. Then, this interface type will be affected by the inherited constraints of its supertypes, and it will be checked with all the other siblings. Thus, adding supertypes to the affected interfaces list is not needed here.

However, if it's deleted, we have to mark the supertypes as affected. When we remove a capability, it is no more counted for the super interface types' cardinalities, and if it is, say, `card(1..)` and abstract, we should return an error.